cpp/src/binaryop/compiled/struct_binary_ops.cuh

```diff
@@ -0,0 +1,128 @@
+    thrust::tabulate(
+      rmm::exec_policy(stream),
+      out.begin<bool>(),
+      out.end<bool>(),
+      [optional_iter, is_lhs_scalar, is_rhs_scalar, device_comparator] __device__(size_type i) {
+        return optional_iter[i].has_value() &&
+               device_comparator(cudf::experimental::row::lhs_index_type{is_lhs_scalar ? 0 : i},
+                                 cudf::experimental::row::rhs_index_type{is_rhs_scalar ? 0 : i});
+      });
+
+  } else {
+    auto device_comparator = table_comparator.less(
+      nullate::DYNAMIC{nullate::DYNAMIC{has_nested_nulls(tlhs) || has_nested_nulls(trhs)}}, c);
+    thrust::tabulate(
+      rmm::exec_policy(stream),
+      out.begin<bool>(),
+      out.end<bool>(),
+      [optional_iter, is_lhs_scalar, is_rhs_scalar, device_comparator] __device__(size_type i) {
+        return optional_iter[i].has_value() &&
+               device_comparator(cudf::experimental::row::lhs_index_type{is_lhs_scalar ? 0 : i},
+                                 cudf::experimental::row::rhs_index_type{is_rhs_scalar ? 0 : i});
+      });
+  }
+  return;
+}
+
+template <typename PhysicalEqualityComparator =
+            cudf::experimental::row::equality::physical_equality_comparator>
+void apply_struct_equality_op(mutable_column_view& out,
+                              column_view const& lhs,
+                              column_view const& rhs,
+                              bool is_lhs_scalar,
+                              bool is_rhs_scalar,
+                              binary_operator op,
+                              PhysicalEqualityComparator c = {},
+                              rmm::cuda_stream_view stream = rmm::cuda_stream_default)
+{
+  CUDF_EXPECTS(op == binary_operator::EQUAL || op == binary_operator::NOT_EQUAL,
+               "Unsupported operator for these types");
+
+  auto tlhs = table_view{{lhs}};
+  auto trhs = table_view{{rhs}};
+  auto table_comparator =
+    cudf::experimental::row::equality::two_table_comparator{tlhs, trhs, stream};
+  auto device_comparator = table_comparator.equal_to(
+    nullate::DYNAMIC{nullate::DYNAMIC{has_nested_nulls(tlhs) || has_nested_nulls(trhs)}},
+    null_equality::EQUAL,
+    c);
+
+  auto outd = column_device_view::create(out, stream);
+  auto optional_iter =
+    cudf::detail::make_optional_iterator<bool>(*outd, nullate::DYNAMIC{out.has_nulls()});
+  thrust::tabulate(
+    rmm::exec_policy(stream),
+    out.begin<bool>(),
+    out.end<bool>(),
+    [optional_iter,
+     is_lhs_scalar,
+     is_rhs_scalar,
+     flip_output = (op == binary_operator::NOT_EQUAL),
+     device_comparator] __device__(size_type i) {
+      auto lhs = cudf::experimental::row::lhs_index_type{is_lhs_scalar ? 0 : i};
+      auto rhs = cudf::experimental::row::rhs_index_type{is_rhs_scalar ? 0 : i};
+      return optional_iter[i].has_value() and
+             (flip_output ? not device_comparator(lhs, rhs) : device_comparator(lhs, rhs));
```

Comment:
This is equivalent:
```suggestion
             (device_comparator(lhs, rhs) == flip_output);
```

Reply:
For this logic, shouldn't it be `(device_comparator(lhs, rhs) != flip_output);`

Reply:
Here's the truth table. It appears to be XOR-like, which is written as `!=` for bools.

|                  | flip true | flip false |
|------------------|-----------|------------|
| comparator true  | false     | true       |
| comparator false | true      | false      |